src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -907,45 +908,36 @@ public static int Sum<T>(this ImmutableArray<T> items, Func<T, int> selector)
 
         internal static void CreateNameToMembersMap
             <TNamespaceOrTypeSymbol, TNamedTypeSymbol, TNamespaceSymbol>
-            (Dictionary<string, object> dictionary, Dictionary<String, ImmutableArray<TNamespaceOrTypeSymbol>> result)
+            (Dictionary<string, object> dictionary, Dictionary<string, ImmutableArray<TNamespaceOrTypeSymbol>> result)
             where TNamespaceOrTypeSymbol : class
             where TNamedTypeSymbol : class, TNamespaceOrTypeSymbol
             where TNamespaceSymbol : class, TNamespaceOrTypeSymbol
         {
-            foreach (var kvp in dictionary)
-            {
-                object value = kvp.Value;
-                ImmutableArray<TNamespaceOrTypeSymbol> members;
+            foreach (var (name, value) in dictionary)
+                result.Add(name, createMembers(value));
 
-                var builder = value as ArrayBuilder<TNamespaceOrTypeSymbol>;
-                if (builder != null)
+            return;
+
+            static ImmutableArray<TNamespaceOrTypeSymbol> createMembers(object value)
+            {
+                if (value is ArrayBuilder<TNamespaceOrTypeSymbol> builder)
                 {
                     Debug.Assert(builder.Count > 1);
-                    bool hasNamespaces = false;
-                    for (int i = 0; i < builder.Count; i++)
+                    foreach (var item in builder)
                     {
-                        if (builder[i] is TNamespaceSymbol)
-                        {
-                            hasNamespaces = true;
-                            break;
-                        }
+                        if (item is TNamespaceSymbol)
+                            return builder.ToImmutableAndFree();
                     }
 
-                    members = hasNamespaces
-                        ? builder.ToImmutable()
```

Comment:
ToImmutable is strictly worse than ToImmutableAndFree.  The latter knows hte builder is being freed, so it can directly move the internal array into the final ImmutableArray (if the counts match).  The former *always* has to copy for teh non-zero element case, which is unnecessary when a move is possible.